cmd/metal-api/internal/datastore/event.go

```diff
@@ -62,7 +79,121 @@ func (rs *RethinkStore) ProvisioningEventForMachine(log *zap.SugaredLogger, even
 	}
 
 	newEC.TrimEvents(100)
-
 	err = rs.UpsertProvisioningEventContainer(newEC)
+	if err != nil {
+		return nil, err
+	}
+
+	if config.Message == states.FSMMessageEnterWaitingState || config.Message == states.FSMMessageLeaveWaitingState {
```

Comment:
Can this be moved into the state machine? I think it increases coupling in the code to have such "after FSM conditions". The main reason to put this here is because was probably because the datastore has the `AdjustNumberOfWaitingMachines` functions, but can't this be passed into the FSM as a function or something like that?